src/lib/operators/jit_operator/jit_operations.hpp

```diff
@@ -0,0 +1,195 @@
+ * A set of generic lambdas defines type-independent versions of these operations. These lambdas can be passed to the
+ * "jit_compute" function to perform the actual computation. The lambdas work on raw, concrete values. It is the
+ * responsibility of "jit_compute" to take care of NULL values, unpack input values and pack the result value.
+ * This way all NULL-value semantics are kept in one place. If either of the inputs is NULL, the result of the
+ * computation is also NULL. If neither input is NULL, the computation lambda is called.
+ *
+ * Inside "jit_compute", a switch statement (generated by the preprocessor) dispatches the 36 data-type combinations
+ * and calls the lambda with the appropriately typed parameters. Invalid type combinations (e.g. adding an int32_t
+ * to a std::string) are handled via the SFINAE pattern (substitution failure is not an error): The lambdas fail the
+ * template substitution if they cannot perform their operation on a given combination of input types. In this case,
+ * the "InvalidTypeCatcher" provides a default implementation that throws an exception.
+ *
+ * The generic lambdas can also be passed to the "jit_compute_type" function. The function uses the same dispatching
+ * mechanisms. But instead of executing a computation, it only determines the result type the computation would
+ * have if it were carried out. This functionality is used to determine the type of intermediate values and
+ * computed output columns.
+ *
+ * Logical operators, IsNull and IsNotNull must be handled separately, since their NULL value semantics are
+ * different (i.e. a NULL as either input does not result in the output being NULL as well).
+ */
+
+#define JIT_APPEND_ENUM_NAMESPACE(enum_value) JitDataType::enum_value
+#define JIT_GET_ENUM_VALUE(index, s) JIT_APPEND_ENUM_NAMESPACE(BOOST_PP_TUPLE_ELEM(3, 1, BOOST_PP_SEQ_ELEM(index, s)))
+#define JIT_GET_DATA_TYPE(index, s) BOOST_PP_TUPLE_ELEM(3, 0, BOOST_PP_SEQ_ELEM(index, s))
+
+#define JIT_COMPUTE_CASE(r, types)                                                                                   \
+  case static_cast<uint8_t>(JIT_GET_ENUM_VALUE(0, types)) << 8 | static_cast<uint8_t>(JIT_GET_ENUM_VALUE(1, types)): \
+    catching_func(lhs.as<JIT_GET_DATA_TYPE(0, types)>(), rhs.as<JIT_GET_DATA_TYPE(1, types)>(), result);             \
+    break;
+
+#define JIT_COMPUTE_TYPE_CASE(r, types)                                                                              \
+  case static_cast<uint8_t>(JIT_GET_ENUM_VALUE(0, types)) << 8 | static_cast<uint8_t>(JIT_GET_ENUM_VALUE(1, types)): \
+    return catching_func(JIT_GET_DATA_TYPE(0, types)(), JIT_GET_DATA_TYPE(1, types)());
+
+/* Arithmetic operators */
+const auto jit_addition = [](const auto& a, const auto& b) -> decltype(a + b) { return a + b; };
+const auto jit_subtraction = [](const auto& a, const auto& b) -> decltype(a - b) { return a - b; };
```

Comment:
How is this different from `std::minus<>`?

Reply:
`std::minus` can't be passed as a parameter without specializing it first. Also, it only takes a single template parameter (= both operands and the result are of the same type), but we want to use it for any combination of types.

Reply:
Since c++14, you can. But I missed the part where the types have to be the same.